bip-0000.mediawiki

```diff
@@ -0,0 +1,335 @@
+
+* Collect each ''outpoint'' used as an input to the transaction
+* Let ''outpoints = outpoint<sub>0</sub> || … || outpoint<sub>n</sub>'', sorted by txid and vout, ascending order
+* Let ''outpoints_hash = sha256(outpoints)''
+
+=== Inputs For Shared Secret Derivation ===
+
+While any UTXO with known output scripts can be used to fund the transaction, the sender and receiver MUST use inputs from the following list when deriving the shared secret:
+
+* ''P2TR (key path spends)''
+* ''P2WPKH''
+* ''P2SH-P2WPKH''
+* ''P2PKH''
+
+Inputs with conditional branches or multiple public keys (e.g ''CHECKMULTISIG'') are not included as this introduces malleability and would allow a sender to re-sign with a different set of public keys after the silent payment output has been derived. This is not a concern when the sender controls all of the inputs, but is an issue for CoinJoins and other collaborative protocols, where a malicious participant can participate in deriving the silent payment address with one set of keys and then re-broadcast the transaction with signatures for a different set of public keys. P2TR can have hidden conditional branches (script path), but we work around that as described below.
+
+''' P2TR (key path spends)'''
+
+The sender MUST use the tweaked private key corresponding to the taproot output key (i.e. the tweaked private key for a key path spend). If the taproot output key is not available, the output cannot be used for sending a silent payment. The receiver always uses the taproot output key when scanning, regardless of whether the taproot output is being used in a key path spend or a script path spend<ref name="why_always_output_pubkey">''' Why not skip all taproot script path spends? ''' This causes malleability issues for CoinJoins. If the silent payments protocol skipped taproot script path spends, this would allow an attacker to join a CoinJoin round, participate in deriving the silent payment address using the tweaked private key for a key path spend, and then broadcast their own version of the transaction using the script path spend. If the receiver were to only consider key path spends, they would skip the attacker's script path spend input when deriving the shared secret and not be able to find the funds.</ref>.
+
+The one exception is script path spends that use NUMS point ''H'' as their internal key (as defined in [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs BIP341: Constructing and spending Taproot outputs]), in which case the output will be skipped for the purposes of shared secret derivation<ref name="why_ignore_h">'''Why skip outputs with H as the internal taproot key?''' If use cases get popularized where the taproot key path cannot be used, these outputs can still be included without getting in the way of making a silent payment, provided they specifically use H as their internal taproot key.</ref>.
```

Comment:
Given that BIP-0341 recommends:
> In order to avoid leaking the information that key path spending is not possible it is recommended to pick a fresh integer r in the range 0...n-1 uniformly at random and use H + rG as internal key.

It seems inadvisable to specify `H` directly in this BIP.

Reply:
There are certain use cases today where only the script path will be used (e.g. inscriptions), where it would be fine (and desirable) to use H as the internal public key. Seeing H clearly indicates to the recipient that this input was not used when deriving the silent payment address and allows them to skip the input when doing the shared secret derivation. Using H also allows the sender to fund silent payment transactions with UTXOs that were only ever intended to be spent as a script path spend.

Also, IIUC, the fact that the key path spend was never possible is only revealed at spending time, which for use cases like inscriptions seems totally fine.

Reply:
I'm led to the same direction as on the exclusion of P2TR script path spends - rather than excluding or including inputs based on their construction, can we include or exclude inputs based on how they are spent? If a P2TR input is spent using its keypath then the spender can compute `ax * Bscan`, so that input MUST be included in the silent payment tweak. If a P2TR input is spent using a script path the recipient cannot know whether the spender could compute `ax * Bscan` so that input MUST NOT be included from the silent payment tweak.

Reply:
Doing it this way introduces a malleability issue with CoinJoins, see https://github.com/bitcoin/bips/pull/1458#discussion_r1221885669